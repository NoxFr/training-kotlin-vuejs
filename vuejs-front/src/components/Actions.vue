<template>
  <v-layout row justify-center mb-5>
    <v-dialog v-model="dialog" persistent max-width="500px">
      <v-btn raised small color="beer" slot="activator">{{btnLabel}}</v-btn>
      <v-form lazy-validation ref="form">
        <v-card>
          <v-card-title>
            <span class="headline">My awesome beer</span>
            <v-card-text>
              <v-container grid-list-md>
                <v-layout wrap>
                  <v-flex xs12 sm6 md4>
                    <v-text-field v-model="newBeer.name" :rules="nameRules" label="Name" required></v-text-field>
                  </v-flex>
                  <v-flex xs12 sm6 md4>
                    <v-text-field v-model="newBeer.country" label="Country"></v-text-field>
                  </v-flex>
                </v-layout>
              </v-container>
              <small>*indicates required field</small>
            </v-card-text>
          </v-card-title>
          <v-card-actions>
            <v-spacer></v-spacer>
            <v-btn color="blue darken-1" flat @click.native="dialog = false">Close</v-btn>
            <v-btn color="blue darken-1" flat @click="submit">Save</v-btn>
          </v-card-actions>
        </v-card>
      </v-form>
    </v-dialog>
  </v-layout>
</template>

<script>
import BeerService from '@/services/beer.service'

export default {
  name: 'tkvActions',
  data: () => ({
    dialog: false,
    btnLabel: 'Add a beer',
    nameRules: [
      v => !!v || 'Name is required'
    ],
    newBeer: {
      name: '',
      country: ''
    }
  }),
  methods: {
    submit() {
      if (this.$refs.form.validate()) {
        this.addBeer()
      }
    },
    addBeer () {
      this.dialog = false
      BeerService.add(this.newBeer).then(() => {
        this.$refs.form.reset()
        this.$emit('beers-update')
      })
    }
  }
}
</script>
