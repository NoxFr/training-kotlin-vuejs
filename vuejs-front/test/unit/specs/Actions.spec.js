import { mount, createLocalVue } from '@vue/test-utils'
import Actions from '@/components/Actions'
import flushPromises from 'flush-promises'
import BeerService from '@/services/beer.service'
import Vuetify from 'vuetify';

jest.mock('@/services/beer.service', () => ({
  add: jest.fn()
}))

describe('Actions.vue', () => {
  let wrapper
  beforeEach(() => {
    const localVue = createLocalVue()
    localVue.use(Vuetify);

    jest.clearAllMocks()

    wrapper = mount(Actions, {
      localVue: localVue
    });
  });

  it('should match snapshot', () => {
    // then
    expect(wrapper.element).toBeDefined();
    expect(wrapper.element).toMatchSnapshot();
  })

  it('should add beer and emit', () => {
    // given
    let beer = {
      name: "beer1",
      country: "country1"
    }
    BeerService.add.mockReturnValue(Promise.resolve([]))
    wrapper.setData({
      newBeer: {
        name: 'beer1',
        country: "country1"
      }
    })

    // when
    wrapper.find({ ref: 'form' }).trigger("submit")

    // FIXME : form.vaidate() always false
    // then
    // expect(BeerService.add).toHaveBeenCalledWith(beer)
    //
    // flushPromises().then(() => {
    //   expect(wrapper.emitted('beers-update')).toBeTruthy()
    // })
  })
})
