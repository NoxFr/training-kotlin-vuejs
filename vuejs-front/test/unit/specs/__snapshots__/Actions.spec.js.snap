// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions.vue should match snapshot 1`] = `
<div
  class="layout row justify-center mb-5"
>
  <div
    class="dialog__container"
    style="display: inline-block;"
  >
    <div
      class="dialog__activator"
    >
      <button
        class="btn btn--small beer"
        raised=""
        type="button"
      >
        <div
          class="btn__content"
        >
          Add a beer
        </div>
      </button>
    </div>
    <div
      class="dialog__content"
      style="z-index: 0;"
      tabindex="-1"
    >
      <div
        class="dialog dialog--persistent"
        style="max-width: 500px; display: none;"
      >
         
        <form
          novalidate="novalidate"
        >
          <div
            class="card"
            style="height: auto;"
          >
            <div
              class="card__title"
            >
              <span
                class="headline"
              >
                My awesome beer
              </span>
               
              <div
                class="card__text"
              >
                <div
                  class="container grid-list-md"
                >
                  <div
                    class="layout wrap"
                  >
                    <div
                      class="flex xs12 sm6 md4"
                    >
                      <div
                        class="input-group input-group--required input-group--text-field primary--text"
                      >
                        <label>
                          Name
                        </label>
                        <div
                          class="input-group__input"
                        >
                          <input
                            aria-label="Name"
                            required="required"
                            tabindex="0"
                            type="text"
                          />
                        </div>
                        <div
                          class="input-group__details"
                        >
                          <!---->
                        </div>
                      </div>
                    </div>
                     
                    <div
                      class="flex xs12 sm6 md4"
                    >
                      <div
                        class="input-group input-group--text-field primary--text"
                      >
                        <label>
                          Country
                        </label>
                        <div
                          class="input-group__input"
                        >
                          <input
                            aria-label="Country"
                            tabindex="0"
                            type="text"
                          />
                        </div>
                        <div
                          class="input-group__details"
                        >
                          <!---->
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <small>
                  *indicates required field
                </small>
              </div>
            </div>
             
            <div
              class="card__actions"
            >
              <div
                class="spacer"
              />
               
              <button
                class="btn btn--flat blue--text text--darken-1"
                type="button"
              >
                <div
                  class="btn__content"
                >
                  Close
                </div>
              </button>
               
              <button
                class="btn btn--flat blue--text text--darken-1"
                type="button"
              >
                <div
                  class="btn__content"
                >
                  Save
                </div>
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
